File: 1249-snapshot-array/snapshot-array.cs
public class SnapshotArray
{
    private readonly List<Dictionary<int, int>> _array;
    private int _snapshotId;

    public SnapshotArray(int length)
    {
        _array = new List<Dictionary<int, int>>(length);
        for (var i = 0; i < length; i++)
        {
            _array.Add(new Dictionary<int, int> {[0] = 0});
        }
    }

    public void Set(int index, int val)
    {
        _array[index][_snapshotId] = val;
    }

    public int Snap()
    {
        return _snapshotId++;
    }

    public int Get(int index, int snapId)
    {
        var dictionary = _array[index];
        
        // If snapId is not in the dictionary we should go back to previous snapshot
        while (!dictionary.ContainsKey(snapId))
        {
            --snapId;
        }

        return dictionary[snapId];
    }
}

File: 1249-snapshot-array/snapshot-array.java
import java.util.HashMap;
import java.util.Map;

class SnapshotArray 
{
    private Map<Integer, Integer>[] snapshots;
    private int snapId;

    public SnapshotArray(int length) 
    {
        snapshots = new HashMap[length];
        for (int i = 0; i < length; i++) 
        {
            snapshots[i] = new HashMap<>();
            snapshots[i].put(0, 0); // Initializing with snapId 0 and value 0
        }
        snapId = 0;
    }

    public void set(int index, int val) 
    {
        snapshots[index].put(snapId, val);
    }

    public int snap() 
    {
        snapId++;
        return snapId - 1;
    }

    public int get(int index, int snap_id) 
    {
        // Find the value in the nearest snapshot less than or equal to snap_id
        while (!snapshots[index].containsKey(snap_id)) 
        {
            snap_id--;
        }
        return snapshots[index].get(snap_id);
    }
}
